File: Day_1/code/02_Implemented_linkedlist.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data): data(data), next(nullptr) {}
};

class LinkedList {
private:
    Node* head;

public:
    LinkedList(): head(nullptr) {}

    void insertAtEnd(int data) {
        Node* newNode = new Node(data);
        if (!head) {
            head = newNode;
            return;
        }
        Node* curr = head;
        while (curr->next) curr = curr->next;
        curr->next = newNode;
    }

    void insertAtStart(int data) {
        Node* newNode = new Node(data);
        newNode->next = head;
        head = newNode;
    }

    void insertAtIndex(int index, int data) {
        if (index < 0) {
            cout << "Index cannot be negative." << endl;
            return;
        }

        if (index == 0) {
            insertAtStart(data);
            return;
        }

        Node* curr = head;
        int count = 0;

        while (curr && count < index - 1) {
            curr = curr->next;
            count++;
        }

        if (!curr) {
            cout << "Index out of bounds." << endl;
            return;
        }

        Node* newNode = new Node(data);
        newNode->next = curr->next;
        curr->next = newNode;
    }

    void removeAtIndex(int index) {
        if (index < 0) {
            cout << "Index cannot be negative." << endl;
            return;
        }

        if (!head) {
            cout << "List is empty." << endl;
            return;
        }

        if (index == 0) {
            Node* temp = head;
            head = head->next;
            delete temp;
            return;
        }

        Node* curr = head;
        int count = 0;

        while (curr->next && count < index - 1) {
            curr = curr->next;
            count++;
        }

        if (!curr->next) {
            cout << "Index out of bounds." << endl;
            return;
        }

        Node* temp = curr->next;
        curr->next = curr->next->next;
        delete temp;
    }

    int get(int index) {
        if (index < 0) {
            cout << "Index cannot be negative." << endl;
            return -1;
        }

        Node* curr = head;
        int count = 0;

        while (curr) {
            if (count == index) return curr->data;
            curr = curr->next;
            count++;
        }

        cout << "Index out of bounds." << endl;
        return -1;
    }

    void printList() {
        Node* curr = head;
        while (curr) {
            cout << curr->data;
            if (curr->next) cout << " -> ";
            curr = curr->next;
        }
        cout << endl;
    }
};

int main() {
    LinkedList ll;
    ll.insertAtEnd(10);
    ll.insertAtEnd(20);
    ll.insertAtStart(5);
    ll.insertAtIndex(1, 7);
    ll.printList(); // 5 -> 7 -> 10 -> 20

    ll.removeAtIndex(2);
    ll.printList(); // 5 -> 7 -> 20

    cout << ll.get(1) << endl;  // 7
    cout << ll.get(10) << endl; // Index out of bounds.

    return 0;
}

File: Day_1/code/02_Implemented_linkedlist.py
# Node class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


# LinkedList class
class LinkedList:
    def __init__(self):
        self.head = None

    def insert_at_end(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def insert_at_start(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insert_at_index(self, index, data):
        if index < 0:
            print("Index cannot be negative.")
            return

        if index == 0:
            self.insert_at_start(data)
            return

        new_node = Node(data)
        current = self.head
        count = 0

        while current and count < index - 1:
            current = current.next
            count += 1

        if not current:
            print("Index out of bounds.")
            return

        new_node.next = current.next
        current.next = new_node

    def remove_at_index(self, index):
        if index < 0:
            print("Index cannot be negative.")
            return

        if not self.head:
            print("List is empty.")
            return

        if index == 0:
            self.head = self.head.next
            return

        current = self.head
        count = 0

        while current.next and count < index - 1:
            current = current.next
            count += 1

        if not current.next:
            print("Index out of bounds.")
            return

        current.next = current.next.next

    def get(self, index):
        if index < 0:
            print("Index cannot be negative.")
            return None

        current = self.head
        count = 0

        while current:
            if count == index:
                return current.data
            current = current.next
            count += 1

        print("Index out of bounds.")
        return None

    def print_list(self):
        current = self.head
        elements = []
        while current:
            elements.append(str(current.data))
            current = current.next
        print(" -> ".join(elements))


if __name__ == "__main__":
    ll = LinkedList()
    ll.insert_at_end(10)
    ll.insert_at_end(20)
    ll.insert_at_start(5)
    ll.insert_at_index(1, 7)
    ll.print_list()         # Output: 5 -> 7 -> 10 -> 20

    ll.remove_at_index(2)
    ll.print_list()         # Output: 5 -> 7 -> 20

    print(ll.get(1))        # Output: 7
    print(ll.get(10))       # Output: Index out of bounds.
